ring/replication_set.go

```diff
@@ -243,6 +243,17 @@ func (r ReplicationSet) GetAddressesWithout(exclude string) []string {
 	return addrs
 }
 
+// ZoneCount returns the number of unique zones represented by instances within this replication set.
+func (r *ReplicationSet) ZoneCount() int {
+	zones := map[string]struct{}{}
```

Comment:
I suspect it's faster if you store the unique set of zones as a slice, because we don't expect more than 3 zones (and iterating a slice 3 times over and over should be faster than lookup up a map). Could you add a microbenchmark and try it, please?

Reply:
Good idea, the slice lookup is faster:

```
goos: darwin
goarch: arm64
pkg: github.com/grafana/dskit/ring
                                                           │   map.txt    │              slice.txt               │
                                                           │    sec/op    │    sec/op     vs base                │
ReplicationSetZoneCount/1_instances_per_zone,_1_zones-10     12.83n ±  1%    25.11n ± 0%   +95.71% (p=0.002 n=6)
ReplicationSetZoneCount/1_instances_per_zone,_2_zones-10     24.29n ±  7%    58.20n ± 1%  +139.65% (p=0.002 n=6)
ReplicationSetZoneCount/1_instances_per_zone,_3_zones-10     31.23n ±  4%    97.15n ± 1%  +211.03% (p=0.002 n=6)
ReplicationSetZoneCount/2_instances_per_zone,_1_zones-10     20.98n ±  6%    28.89n ± 0%   +37.67% (p=0.002 n=6)
ReplicationSetZoneCount/2_instances_per_zone,_2_zones-10     42.05n ±  8%    70.84n ± 0%   +68.47% (p=0.002 n=6)
ReplicationSetZoneCount/2_instances_per_zone,_3_zones-10     56.70n ± 11%   119.00n ± 0%  +109.90% (p=0.002 n=6)
ReplicationSetZoneCount/5_instances_per_zone,_1_zones-10     42.93n ± 21%    40.18n ± 0%    -6.42% (p=0.002 n=6)
ReplicationSetZoneCount/5_instances_per_zone,_2_zones-10     88.44n ± 10%   102.60n ± 2%   +16.02% (p=0.002 n=6)
ReplicationSetZoneCount/5_instances_per_zone,_3_zones-10     136.2n ±  2%    183.8n ± 0%   +34.90% (p=0.002 n=6)
ReplicationSetZoneCount/10_instances_per_zone,_1_zones-10    92.34n ±  3%    59.22n ± 0%   -35.87% (p=0.002 n=6)
ReplicationSetZoneCount/10_instances_per_zone,_2_zones-10    201.3n ±  1%    161.4n ± 0%   -19.82% (p=0.002 n=6)
ReplicationSetZoneCount/10_instances_per_zone,_3_zones-10    322.3n ±  3%    297.9n ± 0%    -7.59% (p=0.002 n=6)
ReplicationSetZoneCount/100_instances_per_zone,_1_zones-10   815.9n ±  4%    408.8n ± 0%   -49.90% (p=0.002 n=6)
ReplicationSetZoneCount/100_instances_per_zone,_2_zones-10   1.693µ ±  5%    1.144µ ± 0%   -32.41% (p=0.002 n=6)
ReplicationSetZoneCount/100_instances_per_zone,_3_zones-10   2.777µ ±  7%    2.113µ ± 0%   -23.90% (p=0.002 n=6)
ReplicationSetZoneCount/300_instances_per_zone,_1_zones-10   2.442µ ±  9%    1.172µ ± 0%   -52.01% (p=0.002 n=6)
ReplicationSetZoneCount/300_instances_per_zone,_2_zones-10   5.053µ ± 11%    3.301µ ± 0%   -34.67% (p=0.002 n=6)
ReplicationSetZoneCount/300_instances_per_zone,_3_zones-10   7.953µ ±  2%    6.135µ ± 1%   -22.86% (p=0.002 n=6)
geomean                                                      208.8n          222.2n         +6.45%

                                                           │   map.txt   │          slice.txt          │
                                                           │    B/op     │    B/op     vs base         │
ReplicationSetZoneCount/1_instances_per_zone,_1_zones-10     0.00 ± 0%     16.00 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/1_instances_per_zone,_2_zones-10     0.00 ± 0%     48.00 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/1_instances_per_zone,_3_zones-10      0.0 ± 0%     112.0 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/2_instances_per_zone,_1_zones-10     0.00 ± 0%     16.00 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/2_instances_per_zone,_2_zones-10     0.00 ± 0%     48.00 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/2_instances_per_zone,_3_zones-10      0.0 ± 0%     112.0 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/5_instances_per_zone,_1_zones-10     0.00 ± 0%     16.00 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/5_instances_per_zone,_2_zones-10     0.00 ± 0%     48.00 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/5_instances_per_zone,_3_zones-10      0.0 ± 0%     112.0 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/10_instances_per_zone,_1_zones-10    0.00 ± 0%     16.00 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/10_instances_per_zone,_2_zones-10    0.00 ± 0%     48.00 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/10_instances_per_zone,_3_zones-10     0.0 ± 0%     112.0 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/100_instances_per_zone,_1_zones-10   0.00 ± 0%     16.00 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/100_instances_per_zone,_2_zones-10   0.00 ± 0%     48.00 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/100_instances_per_zone,_3_zones-10    0.0 ± 0%     112.0 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/300_instances_per_zone,_1_zones-10   0.00 ± 0%     16.00 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/300_instances_per_zone,_2_zones-10   0.00 ± 0%     48.00 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/300_instances_per_zone,_3_zones-10    0.0 ± 0%     112.0 ± 0%  ? (p=0.002 n=6)
geomean                                                                ¹   44.14       ?
¹ summaries must be >0 to compute geomean

                                                           │   map.txt    │          slice.txt          │
                                                           │  allocs/op   │ allocs/op   vs base         │
ReplicationSetZoneCount/1_instances_per_zone,_1_zones-10     0.000 ± 0%     1.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/1_instances_per_zone,_2_zones-10     0.000 ± 0%     2.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/1_instances_per_zone,_3_zones-10     0.000 ± 0%     3.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/2_instances_per_zone,_1_zones-10     0.000 ± 0%     1.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/2_instances_per_zone,_2_zones-10     0.000 ± 0%     2.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/2_instances_per_zone,_3_zones-10     0.000 ± 0%     3.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/5_instances_per_zone,_1_zones-10     0.000 ± 0%     1.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/5_instances_per_zone,_2_zones-10     0.000 ± 0%     2.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/5_instances_per_zone,_3_zones-10     0.000 ± 0%     3.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/10_instances_per_zone,_1_zones-10    0.000 ± 0%     1.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/10_instances_per_zone,_2_zones-10    0.000 ± 0%     2.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/10_instances_per_zone,_3_zones-10    0.000 ± 0%     3.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/100_instances_per_zone,_1_zones-10   0.000 ± 0%     1.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/100_instances_per_zone,_2_zones-10   0.000 ± 0%     2.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/100_instances_per_zone,_3_zones-10   0.000 ± 0%     3.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/300_instances_per_zone,_1_zones-10   0.000 ± 0%     1.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/300_instances_per_zone,_2_zones-10   0.000 ± 0%     2.000 ± 0%  ? (p=0.002 n=6)
ReplicationSetZoneCount/300_instances_per_zone,_3_zones-10   0.000 ± 0%     3.000 ± 0%  ? (p=0.002 n=6)
geomean                                                                 ¹   1.817       ?
¹ summaries must be >0 to compute geomean
```

Reply:
Is it worth the complexity of passing the zone count through from places that create `ReplicationSet`s that know how many zones there are? eg. in [`Ring.Get()`](https://github.com/grafana/dskit/blob/ec568ef3c9f30d6d077c2f7eb960a083dc49217b/ring/ring.go#L349) we know the number of zones in the ring if zone awareness is enabled